src/gui/widgets/tab_container.cpp

```diff
@@ -0,0 +1,198 @@
+/*
+	Copyright (C) 2024
+	by Subhraman Sarkar (babaissarkar) <suvrax@gmail.com>
+	Part of the Battle for Wesnoth Project https://www.wesnoth.org/
+
+	This program is free software; you can redistribute it and/or modify
+	it under the terms of the GNU General Public License as published by
+	the Free Software Foundation; either version 2 of the License, or
+	(at your option) any later version.
+	This program is distributed in the hope that it will be useful,
+	but WITHOUT ANY WARRANTY.
+
+	See the COPYING file for more details.
+*/
+
+#define GETTEXT_DOMAIN "wesnoth-lib"
+
+#include "gui/widgets/tab_container.hpp"
+
+#include "gui/core/log.hpp"
+#include "gettext.hpp"
+#include "gui/auxiliary/find_widget.hpp"
+#include "gui/core/window_builder/helper.hpp"
+#include "gui/core/register_widget.hpp"
+#include "gui/widgets/listbox.hpp"
+#include "gui/widgets/stacked_widget.hpp"
```

Comment:
This header isn't used.

tab_container and stacked_widget seem to be doing similar things. tab_container seems to be simpler, which is good, but is there a reason why there are two implementations? Could we later, but not in this PR, switch the preferences dialog to use tab_container?

Reply:
What are you talking about…? The stacked widget is an "overlay layout" that stacks several grids physically on top of each other, while the tab container is a combined "listbox" and "multi-page". They're completely different.

I don't think there's any reason to drop either the stacked widget or the multi page. They're basic fundamental layout containers, while the tabbed container is a higher-level layout container that combines two of them into one convenient package. So, presumably the multi-page (or _maybe_ the stacked widget) is an important implementation detail of the tabbed container and can't just be replaced with it.

Reply:
> Could we later, but not in this PR, switch the preferences dialog to use tab_container?

Yes, both the preferences and the unit type editor can benefit.

I agree with @CelticMinstrel  that `stacked_widget` and `tab_container` are supposed to do different things. As another example, tab_container is supposed to be similar to Java Swing's [JTabbedPane](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/JTabbedPane.html) or Gtk's [Notebook](https://docs.gtk.org/gtk3/class.Notebook.html), while stacked_widget is similar to Swing's [JLayeredPane](https://docs.oracle.com/javase/7/docs/api/javax/swing/JLayeredPane.html) or probably Gtk's [Overlay](https://docs.gtk.org/gtk4/class.Overlay.html)

The tab container is inspired from [here](https://wiki.wesnoth.org/NotSoEasyCoding#Generic_tab_container_widget). I think the current implementation uses `stacked_widget` only because there was no `tab_container` at that moment.

Reply:
BTW, the `tab_container` here supports only vertical tabs. I have plans to add horizontal tab support to it so that it can be used in some other places, like the Build Info/About dialog and the MP create game dialog.

Reply:
I've removed the header.

Reply:
> As another example, tab_container is supposed to be similar to Java Swing's [JTabbedPane](https://docs.oracle.com/en/java/javase/17/docs/api/java.desktop/javax/swing/JTabbedPane.html) or Gtk's [Notebook](https://docs.gtk.org/gtk3/class.Notebook.html), while stacked_widget is similar to Swing's [JLayeredPane](https://docs.oracle.com/javase/7/docs/api/javax/swing/JLayeredPane.html) or probably Gtk's [Overlay](https://docs.gtk.org/gtk4/class.Overlay.html)

In Slate the multi-page would be SWidgetSwitcher and the stacked widget would be SOverlay. I think there's an equivalent to the tabbed pane as well (with more functionality than this one, supporting dragging tabs around, dragging them _out_, redocking them, etc) but I'm not sure what it's called.

JLayeredPane does look like the same basic thing that the stacked widget does.